src/Clapi/Relay.hs

```diff
@@ -178,45 +206,58 @@ handleTrcud
   -> (OutboundClientUpdateDigest, OutboundProviderDigest)
 handleTrcud vs trcud@(TrcUpdateDigest {trcudNamespace = ns}) =
   let
-    (errMap, opd) = processTrcUpdateDigest vs trcud
+    (errMap, ProtoFrpDigest dat cr cont ) = processTrcUpdateDigest vs trcud
     ocud = (frcudEmpty ns) {frcudErrors = fmap (Text.pack . show) <$> errMap}
   in
-    (ocud, opd)
+    (ocud, FrpDigest ns dat cr cont)
 
 relay
-  :: Monad m => Valuespace
+  :: Monad m => Map Namespace Valuespace
   -> Protocol (i, PostNstInboundDigest) Void (i, OutboundDigest) Void m ()
-relay vs = waitThenFwdOnly fwd
+relay vsm = waitThenFwdOnly fwd
   where
     fwd (i, dig) = case dig of
         PnidRootGet -> do
             sendRev
               ( i
               , Ocrid $ FrcRootDigest $ Map.fromSet (const $ SoAfter Nothing) $
-                Set.fromList $ treeChildNames $ vsTree vs)
-            relay vs
-        PnidCgd cgd ->
-          let (ocsed, ocids) = genInitDigests cgd vs in do
+                Map.keysSet vsm)
+            relay vsm
+        PnidClientGet cgd ->
+          let (ocsed, ocids) = genInitDigests cgd vsm in do
             unless (ocsedNull ocsed) $ sendRev (i, Ocsed ocsed)
             mapM_ (sendRev . (i,) . Ocid) ocids
-            relay vs
-        PnidTrcud trcud ->
-          let (ocud, opd) = handleTrcud vs trcud in do
+            relay vsm
+        PnidTrcud trcud -> let ns = trcudNamespace trcud in maybe
+          (do
+            sendRev (i, Ocud $ (frcudEmpty ns)
+              -- FIXME: Not a global error, NamespaceError?
+              { frcudErrors = Map.singleton GlobalError
+                ["fixthis: Bad namespace"]
+              })
+            relay vsm)
+          (\vs -> let (ocud, opd) = handleTrcud vs trcud in do
             sendRev (i, Opd $ opd)
             sendRev (i, Ocud $ ocud)
-            relay vs
-        PnidTrpd trpd -> either
-          terminalErrors
-          (handleOwnerSuccess trpd) $ processToRelayProviderDigest trpd vs
-        PnidTrprd (TrprDigest ns) ->
-          let vs' = vsRelinquish ns vs in do
-            sendRev (i, Ocrd $ generateRootUpdates vs vs')
-            relay vs'
+            relay vsm
+          )
+          (Map.lookup ns vsm)
+        PnidTrpd trpd ->
+          let
+            ns = trpdNamespace trpd
+            vs = Map.findWithDefault
+                (baseValuespace (Tagged $ unNamespace ns) Editable) ns vsm
```

Comment:
We talked about changing this to "root" (not saying this should be part of this PR, just writing it down as a thing to maybe think about).